protocols/netty/src/main/java/org/apache/james/protocols/netty/SslEngineUtil.java

```diff
@@ -0,0 +1,103 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.protocols.netty;
+
+import org.jboss.netty.channel.Channel;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import java.net.InetSocketAddress;
+
+public enum SslEngineUtil {
+    INSTANCE;
```

Comment:
We usually use Singleton only for immutable things (and it looks like the class is mutable here).
What I suggest is to declare 5 static factories like:
```java
public static SslEngineUtil remoteOnly();
```
First, it allows to have a single instance if we think it worth limiting the number of instance.
Second, all instances will share the same type. We will be able to pass it to our code as a dependency and choose the implementation in the right place.
Third, there's no global shared mutable anymore (and we would be happy about that).

WDYT ?

If you need a hand, I can try to write the solution I have in mind tomorrow, just let me know.

Reply:
Yeah I'd like to see what you have in mind so I can better understand the coding practices you guys use.